doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,549 @@
+guidelines. This list is not exhaustive, and there are always exceptions to
+the rules.*
+
+```{admonition} Public API
+- Function names.
+- Keyword argument names.
+- Keyword argument default values.
+- Submodule names.
+- The mathematical conventions used to define a function.
+```
+
+And here are some things that generally aren't public API, and therefore don't
+require deprecations to change (again, this list is only a general set of
+guidelines).
+
+```{admonition} Not Public API
+- The precise form of an expression. In general, functions may be changed to
+  return a different but mathematically equivalent form of the same
+  expression. This includes a function returning a value which it was not able
+  to compute previously.
+- Positional argument names.
+- Functions and methods that are private, i.e., for internal use only. Such
+  things should generally be prefixed with an underscore `_`, although this
+  convention is not currently universally adhered to in the SymPy codebase.
+- Anything explicitly marked as "experimental".
+- Changes to behavior that were mathematically incorrect previously (in
+  general, bug fixes are not considered breaking changes, because despite the
+  saying, bugs in SymPy are not features).
+- Anything that was added before the most recent release. Code that has not
+  yet made it into a release does not need to be deprecated. If you are going
+  to change the API of new code, it is best to do it before a release is made
+  so that no deprecations are necessary for future releases.
+```
+
+Note: both public and private API functions are included in the [reference
+documentation](reference), and many functions are not included there which
+should be, or are not documented at all which should be, so this should not be
+used to determine whether something public or not.
+
+If you're unsure, there is no harm in deprecating something even if it might
+not actually be "public API".
+
+## The purpose of deprecation
+
+Deprecation has several purposes:
+
+- To allow existing code to continue to work for a while, giving people a
+  chance to upgrade SymPy without fixing all deprecation issues immediately.
+- To warn users that their code will break in a future version.
+- To inform users how to fix their code so that it will continue work in
+  future versions.
+
+All deprecation warnings should be something that users can remove by updating
+their code. Deprecation warnings that fire unconditionally, even when using
+the "correct" newer APIs, should be avoided.
+
+This also means all deprecated code must have a completely functioning
+replacement. If there is no way for users to update their code, then this
+means API in question is not yet read to be deprecated. The deprecation
+warning should inform users of a way to change their code so that it works in
+the same version of SymPy, as well as all future versions (see
+[below](deprecation-documentation)).
```

Comment:
Really it should be possible for users to change their code so that it works in both past and future versions of sympy. In some cases it is possible because the "correct" way of using a particular SymPy feature has not changed and only an "incorrect" way is being deprecated (example: importing something from a submodule rather than using `from sympy import X`). In some cases it isn't possible because the "correct" way doesn't exist in previous sympy versions. Either way there should be at least an attempt to explain how cross-version compatibility can be achieved for any downstream library authors.